bftengine/include/bftengine/ReplicaFactory.hpp

```diff
@@ -0,0 +1,73 @@
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "Replica.hpp"
+#include "ReplicaImp.hpp"
+#include "ReadOnlyReplica.hpp"
+#include "ReplicaLoader.hpp"
+
+namespace preprocessor {
+class PreProcessor;
+}
+
+namespace bftEngine {
+
+class ReplicaFactory {
+ public:
+  ReplicaFactory();
+  ~ReplicaFactory();
+  using IReplicaPtr = std::unique_ptr<IReplica>;
+  IReplicaPtr createNewReplica(const ReplicaConfig &,
+                               std::shared_ptr<IRequestsHandler>,
+                               IStateTransfer *,
+                               bft::communication::ICommunication *,
+                               MetadataStorage *,
+                               std::shared_ptr<concord::performance::PerformanceManager> pm,
+                               const std::shared_ptr<concord::secretsmanager::ISecretsManagerImpl> &sm,
+                               const std::function<void(bool)> &);
+  IReplicaPtr createNewReplica(const ReplicaConfig &,
+                               std::shared_ptr<IRequestsHandler>,
+                               IStateTransfer *,
+                               bft::communication::ICommunication *,
+                               MetadataStorage *,
+                               bool &erasedMetadata,
+                               std::shared_ptr<concord::performance::PerformanceManager> pm,
+                               const std::shared_ptr<concord::secretsmanager::ISecretsManagerImpl> &sm,
+                               const std::function<void(bool)> &);
+
+  IReplicaPtr createNewRoReplica(const ReplicaConfig &,
+                                 std::shared_ptr<IRequestsHandler>,
+                                 IStateTransfer *,
+                                 bft::communication::ICommunication *,
+                                 MetadataStorage *);
+
+  void setAggregator(const std::shared_ptr<concordMetrics::Aggregator> &aggregator);
+
+  static logging::Logger &logger() {
+    static logging::Logger logger_ = logging::getLogger("concord.replica_factory");
+    return logger_;
+  }
+
+ private:
+  void addNewPreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
+                          shared_ptr<IncomingMsgsStorage> &incomingMsgsStorage,
+                          shared_ptr<MsgHandlersRegistrator> &msgHandlersRegistrator,
+                          bftEngine::IRequestsHandler &requestsHandler,
+                          InternalReplicaApi &replica,
+                          concordUtil::Timers &timers,
+                          shared_ptr<concord::performance::PerformanceManager> &pm);
+
+  vector<std::unique_ptr<preprocessor::PreProcessor>> preProcessors_;
```

Comment:
@titusen 
My thoughts:
1) I think that the factory should have no variable members, only factory functions. 
It only creates an IReplica and retruns std::unique_ptr<IReplica> (IReplicaPtr).
2) class Replica should hold a single IReplicaPtr and not a vector, since the production should never have a vector. We should never modify a design. 
3) The code in addNewPreProcessor in IReplica::createNewReplica should be renamed and refactored to create a single preprocessor only.

So after 1+2+3 are done, the production code will make sense, and erase any traces of the fact that we might have multiple preprocessors (which doesn't make sense).

Now we need to deal with the test itself - how ot make it work with multiple preprocessors? I suggest to sit together and think about how to modify the test to make it work with multiple preprocessors. One option is to add a new class (inherit) and compile corebft with it to allow multiple pre-processors, but put it in the test folder.

Reply:
done